programs/survival/villager_auto_trader.sc

```diff
@@ -1,27 +1,57 @@
-//Zombie villagers will trade whatever trades they have (cos they keep them now).
+//For gnembon
+//Basically, this app allows you to trade with zombie villagers that a) had trades before converting and b) can hold items
+//This only allows for trades which take emeralds and spit out something else, as it was OP enough already.
+//It also launches items at closest player within 5m radius, regardless of line of sight
 //By: Ghoulboy
 
-__config()->{'scope'->'global','stay_loaded'->true};
+import('math', '_euclidean_sq');
+
+__config()->{'scope'->'global'};
 
 global_trade_speed=120;//Trade every 120 ticks, so u have time to retrieve item. Same as piglin bartering, so seemed fair
 
-__on_tick()->(
-    if(tick_time()%global_trade_speed==2,
-        for(filter(entity_selector('@e[type=zombie_villager]'),nbt(_~'nbt':'HandItems'):'[]':0:'id'=='minecraft:emerald'),
-            hand_count=nbt(_~'nbt':'HandItems'):'[]':0:'Count';
-            if((trade_list = nbt(_~'nbt':'Offers':'Recipes'):'[]')==0,continue());
-            if(type(trade_list)!='list',trade_list=[trade_list]);
-            trade=first(trade_list,
-                _:'buy':'id'=='minecraft:emerald' &&//only emerald trades
-                _:'buyB':'id'=='minecraft:air' &&//don't want second trade item
-                _:'maxUses'!=0 &&//Cos if not it's cheaty, dont work rn
-                (cost=_:'buy':'Count')<=hand_count//Need to have enough emeralds
-            );
-            if(!trade,continue());
-
-            modify(_,'nbt_merge',str('{HandItems:[{id:"minecraft:emerald",Count:%sb}]}',hand_count-cost));
-
-            spawn('item',pos(_),str('{Item:{id:"%s",Count:%db,PickupDelay:0}}',trade:'sell':'id',trade:'sell':'Count'));
+entity_load_handler('zombie_villager', _(e, new) -> _trade(zombie_villager, null));
+
+
+_trade(zombie_villager, trade)->(
+    if(trade==null,//call this again in case the player messes with the zombie villager's nbt, also cos too lazy to shift elsewhere
+        if((trade_list = nbt(zombie_villager~'nbt':'Offers':'Recipes'):'[]')==0, return());//this return call would only be called once, so it's not expensive
+        trade=first(trade_list,
+            _:'buy':'id'=='minecraft:emerald' &&//only emerald trades
+            _:'buyB':'id'=='minecraft:air' &&//don't want second trade item
+            _:'maxUses'!=0 //Cos if not it's cheaty, dont work rn
+        );
+        if(!trade, return())//again, this is a cheap call and only called once
+    );
+
+    if(nbt(zombie_villager~'nbt':'HandItems'):'[]':0:'id'=='minecraft:emerald' && (hand_count=nbt(zombie_villager~'nbt':'HandItems'):'[]':0:'Count') >= ((cost=trade:'buy':'Count')),
+
+        modify(zombie_villager,'nbt_merge',str('{HandItems:[{id:"minecraft:emerald",Count:%sb}]}',hand_count-cost));
```

Comment:
Just out of curiosity: didn't `invenotry_set` work here?

Reply:
no, i'm afraid it didn't